scripts/updateMessages.js

```diff
@@ -1,51 +1,62 @@
-const fetchSpeechToText = async (url) => {
-  // getting translated message from storage.local
-  let savedUrls = await browser.storage.local.get();
-
-  if (savedUrls) {
-    for (let savedUrl in savedUrls) {
-      if (savedUrl === url) {
-        return savedUrls[savedUrl];
-      }
-    }
+/**
+ * A higher order function that provides caching in the browser's local storage.
+ *
+ * Only supports functions with a single argument that can be stored as an object's key
+ * (string, number, symbol, boolean).
+ */
+const withBrowserLocalCache = (fn) => (arg) => {
+  /** @type {Record<any, any>} */
+  const cache = (await browser.storage.local.get()) || {};
+  if (cache[arg]) {
+    return cache[arg];
   }
 
-  // fetching new data
+  const result = await fn(arg);
+  cache[arg] = result;
+  await browser.storage.local.set(cache);
+
+  return result;
+};
+
+const fetchSpeechToText = async (url) => {
   // TODO: fetch sererUrl from heroku server
   const serverUrl = "http://61e2-78-11-176-160.ngrok.io";
 
-  const headers = {
+  const requestOptions = {
     method: "POST",
     headers: { "Content-Type": "application/json" },
-    body: JSON.stringify({ url: url }),
+    body: JSON.stringify({ url }),
   };
 
-  const response = await fetch(serverUrl + "/speech-to-text", headers);
+  const response = await fetch(serverUrl + "/speech-to-text", requestOptions);
 
-  const message = await response.text();
-
-  if (savedUrls) {
-    savedUrls[url] = message;
-    browser.storage.local.set(savedUrls);
-  } else {
-    browser.storage.local.set({ url: message });
-  }
-
-  return message;
+  return response.text();
 };
 
+const memoizedFetchSpeechToText = withBrowserLocalCache(fetchSpeechToText);
+
+// TODO: use `MutationObserver` instead of polling
```

Comment:
The `MutationObserver` could greatly improve performance and IMO is worth using in the final version of the extension :smile: However, maybe performance won't be a problem - I don't know :smile: That's to be checked empirically

Reply:
I tried using `MutationObserver` and it end up:
1. making a ton of useless updates
2. making too little updates (eg. chaning conversation didn't trigger update)

I will give `MutationObserver` another try and see how it ends.

Reply:
Sure, maybe it won't be too much of a problem to do `querySelector` on an interval, who knows :smile: